browser/src/Services/DragAndDrop.tsx

```diff
@@ -0,0 +1,117 @@
+import * as React from "react"
+import * as DND from "react-dnd"
+
+type Render<T> = (props: T) => React.ReactElement<T>
+type OnDrop = (item: any) => object | void
+type IsValidDrop = (item: any) => boolean
+
+// Drop Target ================================================================
+export interface IDroppeable {
+    isOver?: boolean
+    connectDropTarget?: any
+    onDrop: OnDrop
+    isValidDrop: IsValidDrop
+    canDrop?: boolean
+    accepts: string[] | string
+    render: Render<{
+        canDrop: boolean
+        isOver?: boolean
+        connectDropTarget?: DND.DropTargetConnector
+    }>
+}
+interface DroppedProps {
+    onDrop: OnDrop
+    isValidDrop: IsValidDrop
+}
+
+const DropTarget = {
+    drop(dropped: DroppedProps, monitor: DND.DropTargetMonitor) {
+        return dropped.onDrop({ drag: monitor.getItem(), drop: dropped })
+    },
+    canDrop(props: DroppedProps, monitor: DND.DropTargetMonitor) {
+        return props.isValidDrop({ drag: monitor.getItem(), drop: props })
+    },
+}
+
+const DropCollect = (connect: DND.DropTargetConnector, monitor: DND.DropTargetMonitor) => ({
+    connectDropTarget: connect.dropTarget(),
+    isOver: monitor.isOver(),
+    canDrop: monitor.canDrop(),
+})
+
+/**
+ * A component which can have items of a specific matching type dropped onto it
+ */
+@DND.DropTarget<IDroppeable>(({ accepts }) => accepts, DropTarget, DropCollect)
```

Comment:
Cool use of decorators! We havne't leveraged these much..

Reply:
I kind of like them too but here it wasn't so much a choice as how all the documentation and examples are written out, I would have tried without but didn't want to complicate learning the new tool by doing everything differently from the docs 😄

Reply:
👍 Makes sense to me!